dap_prinz_green_jobs/pipeline/green_measures/skills/green_esco_formatting.py

```diff
@@ -1,43 +1,109 @@
 from dap_prinz_green_jobs.getters.data_getters import (
     load_s3_data,
     save_to_s3,
 )
+
 import pandas as pd
 
+from datetime import datetime as date
+
 if __name__ == "__main__":
+    date_stamp = str(date.today().date()).replace("-", "")
+
     # download esco green skills list
     logger.info("downloading esco green skills taxonomy")
     esco_green_skills = load_s3_data(
         BUCKET_NAME,
         "inputs/data/green_skill_lists/esco/greenSkillsCollection_en.csv",
     ).assign(id=lambda x: x["conceptUri"].str.split("/").str[-1])
 
-    # download formatted esco skills list
-    logger.info("formatting esco green skills taxonomy")
-    formatted_esco_skills = load_s3_data(
-        OJO_BUCKET_NAME,
-        "escoe_extension/outputs/data/skill_ner_mapping/esco_data_formatted.csv",
+    logger.info("Formatting the esco green skills taxonomy")
+
+    esco_green_skills["type"] = "preferredLabel"
+
+    # Expand out the alternate labels
+    esco_green_skills["altLabels_expanded"] = esco_green_skills["altLabels"].apply(
+        lambda x: x.split("|") if isinstance(x, str) else None
     )
+    esco_green_skills_exploded = esco_green_skills.explode("altLabels_expanded")
+    esco_green_skills_exploded["type"] = "altLabel"
 
-    formatted_green_esco_skills = formatted_esco_skills[
-        formatted_esco_skills["id"].isin(esco_green_skills["id"])
-    ].reset_index(drop=True)
+    # Add the preferred labels and the alternate together in the format needed for other steps
+    expanded_green_esco = pd.concat(
+        [
+            esco_green_skills[["id", "preferredLabel", "type", "skillType"]].rename(
+                columns={"preferredLabel": "description"}
+            ),
+            esco_green_skills_exploded[
+                ["id", "altLabels_expanded", "type", "skillType"]
+            ].rename(columns={"altLabels_expanded": "description"}),
+        ]
+    )
+    # Don't include blank descriptions
+    expanded_green_esco = expanded_green_esco[
+        pd.notnull(expanded_green_esco["description"])
+    ]
+
+    # The new descriptions often have trailing whitespaces
+    expanded_green_esco["description"] = expanded_green_esco["description"].apply(
+        lambda x: x.strip()
+    )
+
+    # There are times when the preferred label and the altlabel descriptions are the same for the same id, keep the preferred only
+    deduplicated_expanded_green_esco = (
+        expanded_green_esco.groupby(["id", "description"], sort=False)
+        .apply(
+            lambda x: x
+            if "preferredLabel" not in x["type"].tolist()
+            else x.loc[x.type.eq("preferredLabel")]
+        )
+        .reset_index(drop=True)
+    )
+
+    # There are times when the description is the same but the id isn't, we need to just have one, so keep the one
+    # from skillType = "skill/competence" (rather than "knowledge") if possible otherwise pick the first one.
+    # Note this only happens 19 times
```

Comment:
wow this is a really good catch - feels like we should feed back to esco eventually that not all skills have a unique id associated to them!